vmm: Improve snapshot/restore reliability

sboeuf opened the pull request:
By introducing some refactoring of the `DeviceManager`, and by adding a device tree to the DeviceManager, this pull request managed to improve the snapshot/restore mechanism to make it foolproof.
The device tree contains each device along with its resources. Based on these resources, the DeviceManager can restore each device exactly where it was located before the snapshot.

The pull request only takes care of the virtio devices and the virtio-mmio transport layer, a follow up PR will be needed to handle everything related to PCI.

sboeuf commented:
__/!\ /!\ /!\ DO NOT MERGE /!\ /!\ /!\\__
__/!\ /!\ /!\ Depends on https://github.com/cloud-hypervisor/cloud-hypervisor/pull/1072 /!\ /!\ /!\\__

sameo commented on vmm/src/device_manager.rs:
```diff
@@ -2924,12 +2919,37 @@ impl Snapshottable for DeviceManager {
+            }
+
+            // Restore the node
+            if let Some(device) = self.migratable_devices.get(id) {
+                debug!("Restoring {} from DeviceManager", id);
+                if let Some(snapshot) = snapshot.snapshots.get(id) {
+                    device.lock().unwrap().restore(*snapshot.clone())?;
+                } else {
+                    return Err(MigratableError::Restore(anyhow!("Missing device {}", id)));
+                }
+            }
+
```
As a follow-up improvement, I think we want to have each and every of our device to implement the `Migratable` trait and the device tree `Node` to carry the device as `dyn Migratable` and unconditionally restore each device. Those who are practically not migratable will have an empty implementation for that trait.

For now, we could easily have an `migratable: Option<Arc<Mutex<dyn Migratable>>>>` added to our `Node`, and use our device tree instead of the `migratable_devices` map.

sboeuf commented:
@sameo I think I have addressed all of your comments. Please have a look!
One thing, I could add one more commit in order to add some useful functions to the `DeviceTree`, like `parent()` which would return directly the parent `DeviceNode`.

sameo commented on vmm/src/device_manager.rs:
```diff
@@ -2016,7 +2086,22 @@ impl DeviceManager {
+            // existing entry.
+            self.device_tree.insert(
+                id.clone(),
+                DeviceNode::new(
+                    id,
+                    Some(Arc::clone(&virtio_mem_device) as Arc<Mutex<dyn Migratable>>),
```
It could be worth adding a macro for that, that would take and id and a device as its parameters. It would not bring anything but readability to the code.

sboeuf replied:
I might be missing how to do it correctly because I don't see how this would help the readability. From what I understand, if I was creating a macro `new_device_node!()`, each node creation would look very much the same:
```rust
new_device_node!(id.clone(), Arc::clone(&virtio_mem_device) as Arc<Mutex<dyn Migratable>>)
```
Again, I might be missing your point cause I'm not great with macros :)

sameo replied:
I was thinking about something that would look like:

```Patch
diff --git a/vmm/src/device_manager.rs b/vmm/src/device_manager.rs
index bb30c885..1b946335 100644
--- a/vmm/src/device_manager.rs
+++ b/vmm/src/device_manager.rs
@@ -581,6 +581,15 @@ impl DeviceNode {
     }
 }
 
+macro_rules! device_node {
+    ($id:ident, $device:ident) => {
+        DeviceNode::new(
+            $id.clone(),
+            Some(Arc::clone(&$device) as Arc<Mutex<dyn Migratable>>),
+        )
+    };
+}
+
 #[derive(Clone, Serialize, Deserialize)]
 struct DeviceTree(HashMap<String, DeviceNode>);
 
@@ -953,13 +962,8 @@ impl DeviceManager {
                 // Fill the device tree with a new node. In case of restore, we
                 // know there is nothing to do, so we can simply override the
                 // existing entry.
-                self.device_tree.insert(
-                    iommu_id.clone(),
-                    DeviceNode::new(
-                        iommu_id.clone(),
-                        Some(Arc::clone(&device) as Arc<Mutex<dyn Migratable>>),
-                    ),
-                );
+                self.device_tree
+                    .insert(iommu_id.clone(), device_node!(iommu_id, device));
 
                 (Some(device), Some(mapping))
             } else {
@@ -1578,13 +1582,8 @@ impl DeviceManager {
             // Fill the device tree with a new node. In case of restore, we
             // know there is nothing to do, so we can simply override the
             // existing entry.
-            self.device_tree.insert(
-                id.clone(),
-                DeviceNode::new(
-                    id.clone(),
-                    Some(Arc::clone(&vhost_user_net_device) as Arc<Mutex<dyn Migratable>>),
-                ),
-            );
+            self.device_tree
+                .insert(id.clone(), device_node!(id, vhost_user_net_device));
```

sboeuf replied:
Ah yes that looks better indeed :)
Also, I can create another macro that would take only an `id` for the cases where we provide the `None`.

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -559,13 +559,25 @@ impl Drop for ActivatedBackend {
+            parent: None,
+            children: Vec::new(),
+        }
+    }
```
No because you need the `DeviceNode` structure to implement the `Default` trait for that. And basically the Default trait would do the same thing, but also we would have to pick a default value for `id`, which is why I'd rather not implement the `Default` trait.

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -578,9 +578,31 @@ impl DeviceNode {
+    fn insert(&mut self, k: String, v: DeviceNode) -> Option<DeviceNode> {
+        self.0.insert(k, v)
+    }
+    #[cfg(feature = "pci_support")]
```
Yes, otherwise `clippy` complains :(

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -736,7 +758,7 @@ impl DeviceManager {
             #[cfg(feature = "pci_support")]
             pci_devices: HashMap::new(),
-            device_tree: HashMap::new(),
+            device_tree: DeviceTree(HashMap::new()),
```
Oh yes I missed this one!

sboeuf commented on vmm/src/device_manager.rs:
```diff
@@ -993,16 +1002,14 @@ impl DeviceManager {
+                APIC_START,
+                Arc::clone(&self.msi_interrupt_manager),
+            )
+            .map_err(DeviceManagerError::CreateIoapic)?,
```
Yes it is because in this commit I had to use `id.clone()` instead of just `id`, which triggered `cargo fmt` to reformat the code.

sboeuf commented:
@rbradford @sameo the PR is ready to be fully reviewed. I'd like to make sure we merge it before I go further implementing PCI and virtio-pci for snapshot/restore.

sameo commented on vmm/src/device_manager.rs:
```diff
@@ -611,47 +611,63 @@ impl DeviceTree {
+            let pos = nodes.len();
+            nodes.extend(next_node_layer);
+
+            node_layer = &nodes[pos..];
```
Maybe we could merge that commit with the previous one?

sboeuf replied:
Yes I'll do that.

sboeuf commented:
@sameo everything's addressed now :)

rbradford commented on vm-device/src/lib.rs:
```diff
@@ -7,6 +9,39 @@ use vm_memory::{
+        size: u32,
+    },
+    /// Network Interface Card MAC address.
+    MacAddresss(String),
```
nit: s/Addresss/Address/

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -314,6 +316,21 @@ pub enum DeviceManagerError {
+    MmioRangeAllocation,
+
+    /// Resource was already found.
+    ResourceAlreadyExist,
```
nit: s/ResourceAlreadyExist/ResourceAlreadyExists/

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -542,6 +559,133 @@ impl Drop for ActivatedBackend {
+    device_tree: DeviceTree,
+    device_id_cnt: Wrapping<usize>,
+}
+
```
Could we move the `DeviceTree` into its own module to reduce the length of this file? Along with some unit tests.

sboeuf replied:
Yes I can do that in a new commit, because rebasing is going to be painful otherwise.

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -723,62 +834,111 @@ impl DeviceManager {
+        Ok(())
+    }
+
+    fn state(&self) -> DeviceManagerState {
```
Could we embed `DeviceManagerState` into `DeviceManager` rather than the individual values?

sboeuf replied:
This is gonna create a new level of indirection inside the device manager to access the device tree and the device id counter.
This `DeviceManagerState` structure is not really meant to be maintained at all time, the point is just to create it when we need to save the current state.

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -1219,9 +1397,11 @@ impl DeviceManager {
+            // Fill the device tree with a new node. In case of restore, we
+            // know there is nothing to do, so we can simply override the
+            // existing entry.
+            self.device_tree
```
I'm not sure about all these duplicated comments. Not sure it adds anything to copy and paste everywhere.

sboeuf replied:
Ok I'll remove it.

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -1472,22 +1670,79 @@ impl DeviceManager {
+                        cache_range = Some((*base, *size));
+                    }
+                    _ => {
+                        error!("Unexpected resource for {}", id);
```
Might as well print out the bad resource too.

sboeuf replied:
Done

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -1575,6 +1826,38 @@ impl DeviceManager {
+                        region_range = Some((*base, *size));
+                    }
+                    _ => {
+                        error!("Unexpected resource for {}", id);
```
Ditto.

sboeuf replied:
Done

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -2043,28 +2380,104 @@ impl DeviceManager {
+                        mmio_irq = Some(*irq);
+                    }
+                    _ => {
+                        error!("Unexpected resource for {}", id);
```
Ditto.

sboeuf replied:
What is this ditto for?

sboeuf replied:
Nevermind, I got it.

sboeuf replied:
Done

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -2079,17 +2492,24 @@ impl DeviceManager {
             "virtio_mmio.device={}K@0x{:08x}:{}",
-            MMIO_LEN / 1024,
-            mmio_base.0,
+            mmio_size / 1024,
```
Are we absolutely sure this is a multiple of 1KiB now?

sboeuf replied:
Yes we are since everything (store/restore of resources) is done by this function.

sboeuf commented:
@rbradford comments answered and addressed, PTAL
